src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -75,6 +88,81 @@ unsigned JvmtiTagMapKey::get_hash(const JvmtiTagMapKey& entry) {
   }
 }
 
+static bool equals_oops(oop obj1, oop obj2); // forward declaration
+
+static bool equals_fields(char type, oop obj1, oop obj2, int offset) {
+  switch (type) {
+  case JVM_SIGNATURE_BOOLEAN:
+    return obj1->bool_field(offset) == obj2->bool_field(offset);
+  case JVM_SIGNATURE_CHAR:
+    return obj1->char_field(offset) == obj2->char_field(offset);
+  case JVM_SIGNATURE_FLOAT:
+    return obj1->float_field(offset) == obj2->float_field(offset);
+  case JVM_SIGNATURE_DOUBLE:
+    return obj1->double_field(offset) == obj2->double_field(offset);
+  case JVM_SIGNATURE_BYTE:
+    return obj1->byte_field(offset) == obj2->byte_field(offset);
+  case JVM_SIGNATURE_SHORT:
+    return obj1->short_field(offset) == obj2->short_field(offset);
+  case JVM_SIGNATURE_INT:
+    return obj1->int_field(offset) == obj2->int_field(offset);
+  case JVM_SIGNATURE_LONG:
+    return obj1->long_field(offset) == obj2->long_field(offset);
+  case JVM_SIGNATURE_CLASS:
+  case JVM_SIGNATURE_ARRAY:
+    return equals_oops(obj1->obj_field(offset), obj2->obj_field(offset));
+  }
+  ShouldNotReachHere();
+}
+
+// For heap-allocated objects offset is 0 and 'klass' is obj1->klass() (== obj2->klass()).
+// For flattened objects offset is the offset in the holder object, 'klass' is inlined object class.
+static bool equals_value_objects(oop obj1, oop obj2, InlineKlass* klass, int offset) {
+  for (JavaFieldStream fld(klass); !fld.done(); fld.next()) {
+    // ignore static fields
+    if (fld.access_flags().is_static()) {
+      continue;
+    }
+    int field_offset = offset + fld.offset() - (offset > 0 ? klass->payload_offset() : 0);
+    if (fld.is_flat()) { // flat value field
+      InstanceKlass* holder_klass = fld.field_holder();
+      InlineKlass* field_klass = holder_klass->get_inline_type_field_klass(fld.index());
+      if (!equals_value_objects(obj1, obj2, field_klass, field_offset)) {
+        return false;
+      }
+    } else {
+      if (!equals_fields(fld.signature()->char_at(0), obj1, obj2, field_offset)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+static bool equals_oops(oop obj1, oop obj2) {
+  if (obj1 == obj2) {
+    return true;
+  }
+
+  if (EnableValhalla) {
+    if (obj1 != nullptr && obj2 != nullptr && obj1->klass() == obj2->klass() && obj1->klass()->is_inline_klass()) {
```

Comment:
Nit: I guess, the function `is_inline_type()` can be used instead of `klass()->is_inline_klass()`.